plugins/woocommerce/src/Internal/Traits/RestApiCache.php

```diff
@@ -0,0 +1,405 @@
+ *   in a call to 'with_cache' will be cached using wp_cache_* functions.
+ * - For the purposes of caching, a request is uniquely identified by its route,
+ *   query string, and user ID.
+ * - The VersionStringGenerator class is used to track versions of entities included
+ *   in the responses (an "entity" is any object that is uniquely identified by type and id
+ *   and contributes with information to be included in the response),
+ *   so that when those entities change, the relevant cached responses become invalid.
+ *   Modification of entity versions must be done externally by the code that modifies
+ *   those entities (via calls to VersionStringGenerator::generate_version).
+ * - Caching can be disabled for a given request by adding a '_skip_cache=true|1'
+ *   to the query string.
+ * - A X-WC-Cache HTTP header is added to responses to indicate cache status:
+ *   HIT, MISS, or SKIP.
+ * - Cached response TTL is fixed to one hour.
+ *
+ * Usage: Wrap endpoint callbacks with the `with_cache()` method when registering routes.
+ *
+ * Example:
+ *
+ * class WC_REST_Products_Controller extends WC_REST_Products_V2_Controller {
+ *     use RestApiCache;  // REQUIRED
+ *
+ *     public function __construct() {
+ *         parent::__construct();
+ *         $this->initialize_rest_api_cache();  // REQUIRED
+ *     }
+ *
+ *     protected function get_default_entity_type(): ?string {
+ *         return 'product';  // REQUIRED (or specify entity_type in each with_cache call)
+ *     }
+ *
+ *     public function register_routes() {
+ *         register_rest_route(
+ *             $this->namespace,
+ *             '/' . $this->rest_base . '/(?P<id>[\d]+)',
+ *             array(
+ *                 'methods'  => WP_REST_Server::READABLE,
+ *                 'callback' => $this->with_cache(
+ *                     array( $this, 'get_item' ),
+ *                     array(
+ *                         // String, optional if get_default_entity_type() is overridden.
+ *                         'entity_type' => 'product',
+ *                     )
+ *                 ),
+ *             )
+ *         );
+ *     }
+ * }
+ *
+ * Cache invalidation happens when:
+ * - Entity versions change (tracked via VersionStringGenerator).
+ * - Cached response TTL expires.
+ *
+ * NOTE: This caching mechanism uses the WordPress cache (wp_cache_* functions).
+ * By default caching is only enabled when an external object cache is enabled
+ * (checked via call to VersionStringGenerator::can_use()), so the cache is persistent
+ * across requests and not just for the current request.
+ *
+ * @since   10.4.0
+ */
+trait RestApiCache {
+	/**
+	 * Cache group name for REST API responses.
+	 *
+	 * @var string
+	 */
+	private static string $cache_group = 'woocommerce_rest_api_cache';
+
+	/**
+	 * Cache TTL in seconds.
+	 *
+	 * @var int
+	 */
+	private static int $cache_ttl = HOUR_IN_SECONDS;
```

Comment:
Should this be configurable in some way?  I can see some endpoints where an hour cache might be on the long side, while others (e.g., products) where an hour is pretty short and updates aren't that frequent.